### PR TITLE
Fix restore from backup with `flatten_nested` and `data_type_default_nullable`

### PR DESCRIPTION
**Pull request by kssenii**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix error `"Table has a different definition: ..."` during restore from a backup if table structure depends on `flatten_nested` or `data_type_default_nullable`.

### CODE REVIEW

**Comment by robot-ch-test-poll2** [comment]
*This is an automated comment for commit ea5796fcb6860d7bb1854f8ad39ab7573941cb87 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/56306/ea5796fcb6860d7bb1854f8ad39ab7573941cb87/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/6747405955">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/56306/ea5796fcb6860d7bb1854f8ad39ab7573941cb87/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

File: tests/queries/0_stateless/02907_backup_restore_flatten_nested.sh [hunk 1]
```diff
@@ -0,0 +1,22 @@
+#!/usr/bin/env bash
+
+CUR_DIR=$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)
+# shellcheck source=../shell_config.sh
+. "$CUR_DIR"/../shell_config.sh
+
+${CLICKHOUSE_CLIENT} -nm --query "
+drop table if exists test;
+set flatten_nested = 0;
+create table test (test Array(Tuple(foo String, bar Float64))) ENGINE = MergeTree() ORDER BY tuple();
```
Comment (evillique) [inline tests/queries/0_stateless/02907_backup_restore_flatten_nested.sh:10]: We will soon only flatten the explicitly stated Nested type: https://github.com/ClickHouse/ClickHouse/pull/56132

Comment (kssenii) [reply tests/queries/0_stateless/02907_backup_restore_flatten_nested.sh:10]: Ok, but the same problem with backup restore can be even with explicitly stated Nested type

Comment (evillique) [reply tests/queries/0_stateless/02907_backup_restore_flatten_nested.sh:10]: My suggestion was to change it to something like this for this test to make sense even after the change I mentioned.
```suggestion
create table test (test Nested(foo String, bar Float64)) ENGINE = MergeTree() ORDER BY tuple();
```

File: src/Databases/MySQL/MaterializedMySQLSyncThread.cpp [hunk 2]
```diff
@@ -75,7 +75,7 @@ static BlockIO tryToExecuteQuery(const String & query_to_execute, ContextMutable
         if (!database.empty())
             query_context->setCurrentDatabase(database);
 
-        return executeQuery("/*" + comment + "*/ " + query_to_execute, query_context, true).second;
+        return executeQuery("/*" + comment + "*/ " + query_to_execute, query_context, QueryFlags(true, false)).second;
```
Comment (tavplubix) [inline src/Databases/MySQL/MaterializedMySQLSyncThread.cpp:78]:
```suggestion
        return executeQuery("/*" + comment + "*/ " + query_to_execute, query_context, QueryFlags(/*internal*/ true, /*distributed_backup_restore*/ false)).second;
```

File: src/Databases/DatabaseOnDisk.cpp [hunk 3]
```diff
@@ -73,7 +73,7 @@ std::pair<String, StoragePtr> createTableFromAST(
         auto table_function = factory.get(table_function_ast, context);
         ColumnsDescription columns;
         if (ast_create_query.columns_list && ast_create_query.columns_list->columns)
-            columns = InterpreterCreateQuery::getColumnsDescription(*ast_create_query.columns_list->columns, context, true);
+            columns = InterpreterCreateQuery::getColumnsDescription(*ast_create_query.columns_list->columns, context, true, false);
```
Comment (tavplubix) [inline src/Databases/DatabaseOnDisk.cpp:76]: Same

File: src/Databases/DatabaseReplicated.cpp [hunk 4]
```diff
@@ -916,14 +917,14 @@ void DatabaseReplicated::recoverLostReplica(const ZooKeeperPtr & current_zookeep
         String query = fmt::format("CREATE DATABASE IF NOT EXISTS {} ENGINE=Ordinary", backQuoteIfNeed(to_db_name));
         auto query_context = Context::createCopy(getContext());
         query_context->setSetting("allow_deprecated_database_ordinary", 1);
-        executeQuery(query, query_context, true);
+        executeQuery(query, query_context, QueryFlags(true, false));
```
Comment (tavplubix) [inline src/Databases/DatabaseReplicated.cpp:920]: Same

File: src/Interpreters/InterpreterShowTablesQuery.cpp [hunk 5]
```diff
@@ -214,7 +214,7 @@ BlockIO InterpreterShowTablesQuery::execute()
         return res;
     }
 
-    return executeQuery(getRewrittenQuery(), getContext(), true).second;
+    return executeQuery(getRewrittenQuery(), getContext(), QueryFlags(true, false)).second;
```
Comment (tavplubix) [inline src/Interpreters/InterpreterShowTablesQuery.cpp:217]: Also maybe we can provide default values for the flags, and set only changed ones like `QueryFlags{ .internal = true }`

**Review by tavplubix — APPROVED** [review]

**Comment by tavplubix** [comment]
btw, what about `data_type_default_nullable`?

**Comment by kssenii** [comment]
> btw, what about data_type_default_nullable?

Tested - the same error happens and restore fails. Will fix in a separate PR.

**Comment by kssenii** [comment]
> Will fix in a separate PR.

the fix turned out too easy for a separate PR :)

**Review by tavplubix — APPROVED** [review]

**Comment by kssenii** [comment]
No related failures:

> Integration tests (tsan) [4/6] — fail: 1

test_materialized_mysql_database/test.py::test_utf8mb4

> Stateless tests (release, DatabaseReplicated) [4/4] — fail: 1

02903_rmt_retriable_merge_exception

> Stateless tests (tsan, s3 storage) [1/5] — fail: 1

00993_system_parts_race_condition_drop_zookeeper

**Pull request merged**